### ISSUE TITLE
SAML metadata violates specification

### ISSUE THREAD

**Issue by nijel** [zammad/zammad#3339]
### Infos:

* Used Zammad version: 3.6.x
* Installation method (source, package, ..): zammad.com
* Operating system: any
* Database + version: any
* Elasticsearch version: any
* Browser + version: any

### Expected behavior:

* I expect SAML metadata to follow the [SAML 2.0 specification](https://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf)
* It states: When used as the root element of a metadata instance, this element MUST contain either a validUntil or cacheDuration attribute.

### Actual behavior:

* The metadata lack validity information
* This causes issues with conforming IDP, see https://github.com/OTA-Insight/djangosaml2idp/issues/96

### Steps to reproduce the behavior:

* [Download SAML meta](https://support.zammad.com/auth/saml/metadata) and observe lack of validUntil and cacheDuration attributes on the root element.

Yes I'm sure this is a bug and no feature request or a general question.

**Comment by MrGeneration** [comment]
I see this as an enhancement.
The linked document states on page 9 (point 2.3.1) this:

* validUntil [Optional]
* cacheDuration [Optional]

And: 
```
When used as the root element of a metadata instance, this element MUST contain either a validUntil
or cacheDuration attribute. It is RECOMMENDED that only the root element of a metadata instance
contain either attribute.
```

**Comment by nijel** [comment]
Yes, that's what I quoted. The root element must contain one of these and it is recommended that only the root element contains it.

That's why the attributes themselves are optional because they should appear only on the root element and are not recommended to use elsewhere (the EntityDescriptor element doesn't have to be a root element).

**Comment by thorsteneckel** [comment]
Hey @nijel - I did some research on that topic. Let me share what I found out:

Zammad uses the [`omniauth-saml` gem](https://github.com/omniauth/omniauth-saml) for [providing SAML authentication](https://github.com/zammad/zammad/blob/b85566c53003122ab3c710693a80e8816064bf44/lib/omniauth/saml_database.rb#L1). `omniauth-saml` [utilises the `ruby-saml` gem for generating the Metdata response](https://github.com/omniauth/omniauth-saml/blob/7a7ad49ad425f7ec17d77078bf4bd2dd46e918c5/lib/omniauth/strategies/saml.rb#L246). Which turns out [has no option/way/default](https://github.com/onelogin/ruby-saml/blob/9f710c5028b069bfab4b9e2b66891e0549765af5/lib/onelogin/ruby-saml/metadata.rb) to provide the `validUntil` or `cacheDuration` in the response.
I searched both of those those two repositories and only found occurrences in [an XML schema file in the `ruby-saml` repository](https://github.com/onelogin/ruby-saml/blob/17fce0a6caa6282ebd51c6276d3a7b91138b8d24/lib/schemas/saml-schema-metadata-2.0.xsd#L87-L88) which are both marked as optional. Which is "right" in a way, as you described earlier. However, I guess that this schema format doesn't provide a way to reflect the specification. This constellation causes that their tests will pass even that it's actually not conform with the specification.
[I haven't found a reported issue in the `omniauth-saml` repository](https://github.com/omniauth/omniauth-saml/issues?q=is%3Aissue+validUntil+). I guess the next step would be to ask how the `validUntil` and/or `cacheDuration` Metadata attributes could be set. Would you mind creating an issue?

**Comment by nijel** [comment]
Okay, I did it here: https://github.com/onelogin/ruby-saml/issues/563. It would be great if you could watch that issue because I have no knowledge on how Zammad uses that library or what would be the best solution to this issue for Zammad.

**Comment by thorsteneckel** [comment]
Thanks @nijel ! Watching 👀

**Comment by thorsteneckel** [comment]
Hi @nijel impressed to see that the great people over at `onelogin/ruby-saml` resolved the issue so fast! Now we're only lacking of a way to use this new interface in the https://github.com/omniauth/omniauth-saml gem we're using. Would you mind creating an issue over there as well?
